File: ios/RTSDA/Services/YouTubeService.swift
import Foundation

// Cache class for API responses
final class YouTubeAPICache: Codable {
    let data: Data
    let timestamp: Date
    
    // Cache API responses for 1 hour
    static let validityDuration: TimeInterval = 60 * 60
    
    init(data: Data, timestamp: Date = Date()) {
        self.data = data
        self.timestamp = timestamp
    }
    
    var isValid: Bool {
        return Date().timeIntervalSince(timestamp) < Self.validityDuration
    }
}

@globalActor
actor YouTubeService {
    static let shared = YouTubeService()
    
    private let configService = ConfigService.shared
    private let channelId = YTConfig.YouTube.channelId
    private let bundleId = Bundle.main.bundleIdentifier ?? "com.rtsda.app"
    
    private lazy var session: URLSession = {
        let config = URLSessionConfiguration.default
        config.httpAdditionalHeaders = [
            "X-Ios-Bundle-Identifier": bundleId,
            "User-Agent": "RTSDA-iOS/\(Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "1.0")"
        ]
        return URLSession(configuration: config)
    }()
    
    private var quotaUsageDate: Date?
    private var dailyQuotaUsage: Int = 0
    private let maxDailyQuota = 10000
    private let quotaCostPerVideoLookup = 1  // Adjust this based on actual API cost
    
    private let cache = NSCache<NSString, YouTubeAPICache>()
    private let fileManager = FileManager.default
    private lazy var cacheDirectory: URL? = {
        return fileManager.urls(for: .cachesDirectory, in: .userDomainMask).first?.appendingPathComponent("YouTubeAPICache")
    }()
    
    private func checkQuota() async throws {
        // Reset quota if it's a new day
        if let lastDate = quotaUsageDate, !Calendar.current.isDate(lastDate, inSameDayAs: Date()) {
            quotaUsageDate = Date()
            dailyQuotaUsage = 0
        }
        
        // Initialize date if first time
        if quotaUsageDate == nil {
            quotaUsageDate = Date()
        }
        
        // Check if we have enough quota
        if dailyQuotaUsage + quotaCostPerVideoLookup > maxDailyQuota {
            throw NSError(
                domain: "YouTubeService",
                code: -1,
                userInfo: [NSLocalizedDescriptionKey: "Daily YouTube API quota exceeded. Please try again tomorrow."]
            )
        }
    }
    
    private func incrementQuota() {
        dailyQuotaUsage += quotaCostPerVideoLookup
    }
    
    private init() {
        createCacheDirectoryIfNeeded()
    }
    
    private func createCacheDirectoryIfNeeded() {
        guard let cacheDirectory = cacheDirectory else { return }
        if !fileManager.fileExists(atPath: cacheDirectory.path) {
            try? fileManager.createDirectory(at: cacheDirectory, withIntermediateDirectories: true)
        }
    }
    
    private func getCachedResponse(for url: URL) -> Data? {
        let key = url.absoluteString as NSString
        
        // Check memory cache first
        if let cached = cache.object(forKey: key), cached.isValid {
            print("📦 Using cached response for: \(url.absoluteString)")
            return cached.data
        }
        
        // Check disk cache
        guard let cacheDirectory = cacheDirectory else { return nil }
        let fileURL = cacheDirectory.appendingPathComponent(key.hash.description + ".cache")
        
        guard let data = try? Data(contentsOf: fileURL),
              let cached = try? JSONDecoder().decode(YouTubeAPICache.self, from: data),
              cached.isValid else {
            return nil
        }
        
        // Update memory cache
        cache.setObject(cached, forKey: key)
        print("📦 Using disk-cached response for: \(url.absoluteString)")
        return cached.data
    }
    
    private func cacheResponse(_ data: Data, for url: URL) {
        let key = url.absoluteString as NSString
        let cache = YouTubeAPICache(data: data)
        
        // Save to memory cache
        self.cache.setObject(cache, forKey: key)
        
        // Save to disk cache
        guard let cacheDirectory = cacheDirectory else { return }
        let fileURL = cacheDirectory.appendingPathComponent(key.hash.description + ".cache")
        
        if let encodedData = try? JSONEncoder().encode(cache) {
            try? encodedData.write(to: fileURL)
        }
        print("💾 Cached response for: \(url.absoluteString)")
    }
    
    private func clearExpiredCache() {
        guard let cacheDirectory = cacheDirectory else { return }
        
        do {
            let files = try fileManager.contentsOfDirectory(at: cacheDirectory, includingPropertiesForKeys: nil)
            for file in files {
                guard let data = try? Data(contentsOf: file),
                      let cached = try? JSONDecoder().decode(YouTubeAPICache.self, from: data),
                      !cached.isValid else { continue }
                
                try? fileManager.removeItem(at: file)
                print("🗑️ Removed expired cache file: \(file.lastPathComponent)")
            }
        } catch {
            print("❌ Error clearing cache: \(error)")
        }
    }
    
    private func fetchWithRetry<T: Decodable>(request: URLRequest, retries: Int = 2) async throws -> T {
        var lastError: Error?
        
        for attempt in 0...retries {
            do {
                if attempt > 0 {
                    try await Task.sleep(nanoseconds: UInt64(pow(2.0, Double(attempt))) * 1_000_000_000)
                }
                
                if let cachedResponse = getCachedResponse(for: request.url!) {
                    let response = try JSONDecoder().decode(T.self, from: cachedResponse)
                    return response
                }
                
                let (data, response) = try await session.data(for: request)
                
                if let httpResponse = response as? HTTPURLResponse {
                    if httpResponse.statusCode != 200 {
                        if let errorJson = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                           let error = errorJson["error"] as? [String: Any],
                           let message = error["message"] as? String {
                            print("YouTube API Error: \(message)")
                            throw NSError(domain: "YouTubeAPI", 
                                        code: httpResponse.statusCode,
                                        userInfo: [NSLocalizedDescriptionKey: message])
                        }
                    }
                }
                
                cacheResponse(data, for: request.url!)
                
                return try JSONDecoder().decode(T.self, from: data)
            } catch {
                lastError = error
                print("API Error: \(error)")
                if attempt == retries {
                    throw error
                }
            }
        }
        
        throw lastError!
    }
    
    struct Video: Identifiable {
        let id: String
        let title: String
        let description: String
        let thumbnailURL: String
        let isLiveStream: Bool
        let duration: String?
    }
    
    private func isShortVideo(duration: String?) -> Bool {
        guard let duration = duration else { 
            print("No duration provided, treating as not a short")
            return false 
        }
        
        // Convert ISO 8601 duration to seconds
        let pattern = #"PT(?:(\d+)H)?(?:(\d+)M)?(?:(\d+)S)?"#
        let regex = try! NSRegularExpression(pattern: pattern)
        let range = NSRange(duration.startIndex..<duration.endIndex, in: duration)
        
        guard let match = regex.firstMatch(in: duration, range: range) else { 
            print("Could not parse duration: \(duration)")
            return false 
        }
        
        let hours = match.range(at: 1).location != NSNotFound ? Int(duration[Range(match.range(at: 1), in: duration)!]) ?? 0 : 0
        let minutes = match.range(at: 2).location != NSNotFound ? Int(duration[Range(match.range(at: 2), in: duration)!]) ?? 0 : 0
        let seconds = match.range(at: 3).location != NSNotFound ? Int(duration[Range(match.range(at: 3), in: duration)!]) ?? 0 : 0
        
        let totalSeconds = hours * 3600 + minutes * 60 + seconds
        let isShort = totalSeconds < 61
        print("Video duration: \(hours)h \(minutes)m \(seconds)s (total: \(totalSeconds)s)")
        if isShort {
            print("Video is a Short (duration: \(duration))")
        }
        return isShort
    }
    
    private func isWorshipService(title: String) -> Bool {
        let keywords = ["Worship Service", "Sabbath Service", "Divine Service"]
        let matches = keywords.filter { title.localizedCaseInsensitiveContains($0) }
        if !matches.isEmpty {
            print("Found service keywords in title: \(matches.joined(separator: ", "))")
            return true
        }
        return false
    }
    
    func fetchLatestSermon() async throws -> Video? {
        print("Fetching latest sermon...")
        // Get API key from Remote Config
        let apiKey = await configService.getString(forKey: ConfigService.Keys.youtubeApiKey)
        print("Using channel ID: \(channelId)")
        
        // Clear expired cache entries
        clearExpiredCache()
        
        // First, search for videos
        let searchUrlString = "https://www.googleapis.com/youtube/v3/search?part=snippet&channelId=\(channelId)&maxResults=25&order=date&type=video&key=\(apiKey)"
        guard let searchUrl = URL(string: searchUrlString) else { 
            print("Failed to create search URL")
            return nil 
        }
        
        print("Search URL: \(searchUrlString)")
        
        let searchRequest = URLRequest(url: searchUrl)
        do {
            let searchResponse: SearchResponse = try await fetchWithRetry(request: searchRequest)
            print("Found \(searchResponse.items.count) videos in search")
            
            if searchResponse.items.isEmpty {
                print("No videos found in search response")
                return nil
            }
            
            // Get video details to check duration
            let videoIds = searchResponse.items.map { $0.id.videoId }.joined(separator: ",")
            let detailsUrlString = "https://www.googleapis.com/youtube/v3/videos?part=contentDetails,snippet,status&id=\(videoIds)&key=\(apiKey)"
            guard let detailsUrl = URL(string: detailsUrlString) else { 
                print("Failed to create details URL")
                return nil 
            }
            
            print("Details URL: \(detailsUrlString)")
            
            let detailsRequest = URLRequest(url: detailsUrl)
            let detailsResponse: VideoListResponse = try await fetchWithRetry(request: detailsRequest)
            
            print("Got details for \(detailsResponse.items.count) videos")
            
            // Print all video titles and durations for debugging
            for video in detailsResponse.items {
                print("\nVideo: \(video.snippet.title)")
                print("Duration: \(video.contentDetails.duration)")
                print("Is Short: \(isShortVideo(duration: video.contentDetails.duration))")
                print("Is Worship: \(isWorshipService(title: video.snippet.title))")
                print("LiveBroadcastContent: \(video.snippet.liveBroadcastContent)")
            }
            
            // Find the first video that's not a Short, not a livestream, and not a worship service
            guard let video = detailsResponse.items.first(where: { video in
                print("\nChecking video: \(video.snippet.title)")
                
                let notShort = !isShortVideo(duration: video.contentDetails.duration)
                if !notShort {
                    print("Filtered out: Is a short video")
                    return false
                }
                
                let notLivestream = video.snippet.liveBroadcastContent == "none"
                if !notLivestream {
                    print("Filtered out: Is a livestream (\(video.snippet.liveBroadcastContent))")
                    return false
                }
                
                let notWorshipService = !isWorshipService(title: video.snippet.title)
                if !notWorshipService {
                    print("Filtered out: Is a worship service")
                    return false
                }
                
                print("Video passed all filters")
                return true
            }) else {
                print("No suitable videos found after filtering")
                return nil
            }
            
            print("Selected video: \(video.snippet.title)")
            
            return Video(
                id: video.id,
                title: video.snippet.title,
                description: video.snippet.description,
                thumbnailURL: video.snippet.thumbnails.high?.url ?? video.snippet.thumbnails.default.url,
                isLiveStream: false,
                duration: video.contentDetails.duration
            )
        } catch {
            print("Error fetching latest sermon: \(error)")
            throw error
        }
    }
    
    func fetchUpcomingLivestream() async throws -> Video? {
        print("\nFetching upcoming livestream...")
        let apiKey = await configService.getString(forKey: ConfigService.Keys.youtubeApiKey)
        print("Using channel ID: \(channelId)")
        
        // Clear expired cache entries
        clearExpiredCache()
        
        let searchUrlString = "https://www.googleapis.com/youtube/v3/search?part=snippet&channelId=\(channelId)&eventType=upcoming&type=video&order=date&maxResults=1&key=\(apiKey)"
        guard let searchUrl = URL(string: searchUrlString) else { 
            print("Failed to create search URL")
            return nil 
        }
        
        print("Search URL: \(searchUrlString)")
        
        let request = URLRequest(url: searchUrl)
        do {
            let response: SearchResponse = try await fetchWithRetry(request: request)
            print("Found \(response.items.count) upcoming livestreams")
            
            guard let item = response.items.first else {
                print("No upcoming livestreams found")
                return nil
            }
            
            print("Found upcoming livestream: \(item.snippet.title)")
            
            return Video(
                id: item.id.videoId,
                title: item.snippet.title,
                description: item.snippet.description,
                thumbnailURL: item.snippet.thumbnails.high?.url ?? item.snippet.thumbnails.default.url,
                isLiveStream: true,
                duration: nil
            )
        } catch {
            print("Error fetching upcoming livestream: \(error)")
            throw error
        }
    }
    
    func extractVideoURL(from videoId: String) async throws -> URL {
        // Check cache first
        if let cachedURL = VideoCacheManager.shared.getCachedVideo(id: videoId) {
            print("Using cached video URL for \(videoId)")
            return cachedURL
        }
        
        // Check quota before making API call
        try await checkQuota()
        
        print("Cache miss for \(videoId), fetching from API")
        
        // Use embedded player URL which works with AVPlayer
        let embedURL = "https://www.youtube.com/embed/\(videoId)?playsinline=1&autoplay=1"
        guard let videoURL = URL(string: embedURL) else {
            throw NSError(domain: "YouTubeService", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid video URL"])
        }
        
        // Cache the video URL
        VideoCacheManager.shared.cacheVideo(id: videoId, url: videoURL)
        print("Cached video URL for \(videoId)")
        
        return videoURL
    }
}

// Response models
private struct SearchResponse: Codable {
    let items: [SearchItem]
}

private struct SearchItem: Codable {
    let id: VideoId
    let snippet: Snippet
}

private struct VideoId: Codable {
    let videoId: String
}

private struct Snippet: Codable {
    let title: String
    let description: String
    let thumbnails: Thumbnails
    let liveBroadcastContent: String
}

private struct Thumbnails: Codable {
    let `default`: Thumbnail
    let high: Thumbnail?
}

private struct Thumbnail: Codable {
    let url: String
}

private struct VideoListResponse: Codable {
    let items: [VideoItem]
}

private struct VideoItem: Codable {
    let id: String
    let snippet: Snippet
    let contentDetails: ContentDetails
}

private struct ContentDetails: Codable {
    let duration: String
}

private struct VideoPlayerResponse: Codable {
    let items: [VideoPlayerItem]
}

private struct VideoPlayerItem: Codable {
    let player: PlayerDetails
}

private struct PlayerDetails: Codable {
    let embedHtml: String
}

File: android/app/src/main/java/com/rtsda/appr/service/YouTubeService.kt
package com.rtsda.appr.service

import android.util.Log
import com.google.gson.Gson
import com.rtsda.appr.BuildConfig
import com.rtsda.appr.model.Message
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import java.util.*
import javax.inject.Inject
import javax.inject.Singleton
import java.util.regex.Pattern

@Singleton
class YouTubeService @Inject constructor(
    private val client: OkHttpClient,
    private val gson: Gson
) {
    private val apiKey = BuildConfig.YOUTUBE_API_KEY
    private val channelId = BuildConfig.YOUTUBE_CHANNEL_ID
    private val cacheDuration = 15 * 60 * 1000L // 15 minutes
    
    private var cachedSermon: Message? = null
    private var cachedLivestream: Message? = null
    private var lastFetchTime: Date? = null
    private val mutex = Mutex()
    
    suspend fun getLatestSermon(): Message? = mutex.withLock {
        if (shouldRefreshCache()) {
            refreshCache()
        }
        return cachedSermon
    }
    
    suspend fun getUpcomingLivestream(): Message? = mutex.withLock {
        if (shouldRefreshCache()) {
            refreshCache()
        }
        return cachedLivestream
    }
    
    private fun shouldRefreshCache(): Boolean {
        val now = Date()
        return lastFetchTime?.let { last ->
            now.time - last.time > cacheDuration
        } ?: true
    }
    
    private suspend fun refreshCache() {
        try {
            fetchVideos()
            lastFetchTime = Date()
        } catch (e: Exception) {
            Log.e("YouTubeService", "Error refreshing cache", e)
        }
    }
    
    private suspend fun fetchVideos() {
        try {
            // First search for upcoming livestreams
            val upcomingUrl = buildString {
                append("https://www.googleapis.com/youtube/v3/search")
                append("?key=$apiKey")
                append("&channelId=$channelId")
                append("&part=snippet,id")
                append("&type=video")
                append("&eventType=upcoming")
                append("&maxResults=1")
            }
            
            val upcomingResponse = makeRequest(upcomingUrl)
            val upcomingSearchResponse = gson.fromJson(upcomingResponse, SearchResponse::class.java)
            
            // Set livestream if found
            upcomingSearchResponse.items.firstOrNull()?.let { livestream ->
                Log.d("YouTubeService", "Found upcoming livestream: ${livestream.snippet.title}")
                cachedLivestream = Message(
                    id = livestream.id.videoId,
                    title = livestream.snippet.title,
                    description = livestream.snippet.description,
                    thumbnailUrl = livestream.snippet.thumbnails.high.url,
                    isLivestream = true
                )
            }
            
            // Then search for regular videos (potential sermons)
            val regularUrl = buildString {
                append("https://www.googleapis.com/youtube/v3/search")
                append("?key=$apiKey")
                append("&channelId=$channelId")
                append("&part=snippet,id")
                append("&type=video")
                append("&order=date")
                append("&maxResults=10")
            }
            
            val regularResponse = makeRequest(regularUrl)
            Log.d("YouTubeService", "Regular videos response: $regularResponse")
            val regularSearchResponse = gson.fromJson(regularResponse, SearchResponse::class.java)
            
            val videos = regularSearchResponse.items
            Log.d("YouTubeService", "Found ${videos.size} regular videos")
            
            // Look for most recent sermon
            videos.firstOrNull { video ->
                val title = video.snippet.title.uppercase()
                val description = video.snippet.description.uppercase()
                
                Log.d("YouTubeService", "Checking regular video: ${video.snippet.title}")
                
                // Exclude videos with "worship service" in the title
                val isNotWorshipService = !title.contains("WORSHIP SERVICE") && 
                                        !title.contains("DIVINE SERVICE") &&
                                        !title.contains("DIVINE WORSHIP")
                
                val isSermon = (title.contains("SERMON") || 
                             title.contains("SABBATH") ||
                             title.contains("MESSAGE") ||
                             description.contains("SERMON") ||
                             description.contains("SABBATH MESSAGE"))
                
                val isNotShort = !title.contains("#SHORTS") && !title.contains("SHORT")
                val isNotLive = !title.contains("LIVE") && video.snippet.liveBroadcastContent != "upcoming"
                
                // Check video duration
                val duration = getVideoDuration(video.id.videoId)
                val isLongEnough = duration >= 60
                
                Log.d("YouTubeService", """
                    Video check:
                    Title: ${video.snippet.title}
                    Is sermon: $isSermon
                    Is not worship service: $isNotWorshipService
                    Is not short: $isNotShort
                    Is not live: $isNotLive
                    Duration: ${duration}s
                    Is long enough: $isLongEnough
                """.trimIndent())
                
                isSermon && isNotWorshipService && isNotShort && isNotLive && isLongEnough
            }?.let { sermon ->
                Log.d("YouTubeService", "Found sermon: ${sermon.snippet.title}")
                cachedSermon = Message(
                    id = sermon.id.videoId,
                    title = sermon.snippet.title,
                    description = sermon.snippet.description,
                    thumbnailUrl = sermon.snippet.thumbnails.high.url,
                    isLivestream = false
                )
            } ?: Log.d("YouTubeService", "No sermon found in regular videos")
            
        } catch (e: Exception) {
            Log.e("YouTubeService", "Error fetching videos: ${e.message}")
            e.printStackTrace()
        }
    }
    
    private suspend fun makeRequest(url: String): String = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url(url)
            .build()
            
        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                throw Exception("API call failed with code ${response.code}")
            }
            return@withContext response.body?.string() ?: throw Exception("Empty response body")
        }
    }
    
    data class SearchResponse(
        val items: List<SearchResult>
    )

    data class SearchResult(
        val id: VideoId,
        val snippet: Snippet
    )

    data class VideoId(
        val videoId: String
    )

    data class Snippet(
        val title: String,
        val description: String,
        val thumbnails: Thumbnails,
        val liveBroadcastContent: String
    )

    data class Thumbnails(
        val high: ThumbnailDetails
    )

    data class ThumbnailDetails(
        val url: String
    )

    data class VideoResponse(
        val items: List<VideoItem>
    )
    
    data class VideoItem(
        val contentDetails: ContentDetails
    )
    
    data class ContentDetails(
        val duration: String
    )
    
    private fun parseDuration(duration: String): Long {
        var seconds = 0L
        val pattern = Pattern.compile("PT(?:(\\d+)H)?(?:(\\d+)M)?(?:(\\d+)S)?")
        val matcher = pattern.matcher(duration)
        
        if (matcher.find()) {
            val hours = matcher.group(1)?.toLong() ?: 0
            val minutes = matcher.group(2)?.toLong() ?: 0
            val secs = matcher.group(3)?.toLong() ?: 0
            
            seconds = hours * 3600 + minutes * 60 + secs
        }
        return seconds
    }
    
    private suspend fun getVideoDuration(videoId: String): Long {
        try {
            val url = buildString {
                append("https://www.googleapis.com/youtube/v3/videos")
                append("?key=$apiKey")
                append("&id=$videoId")
                append("&part=contentDetails")
            }
            
            val response = makeRequest(url)
            val videoResponse = gson.fromJson(response, VideoResponse::class.java)
            
            return videoResponse.items.firstOrNull()?.let { video ->
                parseDuration(video.contentDetails.duration)
            } ?: 0
        } catch (e: Exception) {
            Log.e("YouTubeService", "Error getting video duration: ${e.message}")
            return 0
        }
    }
}
